module: better error for named exports from cjs

MylesBorins opened the pull request:
We do not support importing named exports from a CJS module.
This change decorates the error message for missing named exports in
the case that the module being imported is expected to be CJS by the
ESM loader.

Some high level questions

* Are folks ok with how I've decorated the error? It is a bit fragile, as it relies on the format of the error in V8.
* Are folks ok with the copy of the error message? I took a quick stab at it but it could maybe be better

Needs a test, but wanted to confirm folks would be open to doing this before hacking on a test.

guybedford commented:
Definitely a +1 to this, it is much needed.

Ideally the error message could contain the exact context and solution, eg something like:

> No named export named "asdf" for CommonJS module /path/to/module.js
> CommonJS modules only have a "default" export, try importing this instead

mathiasbynens commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -92,6 +93,21 @@ class ModuleJob {
     } catch (e) {
+      if (e.message.includes('does not provide an export named')) {
+        const rootURL = this.module.url;
+        const childSpecifier = e.message.match(/'(.*?)'/)[1];
```
Since `*` already means "0 or more matches", it seems you can drop the `?` here.

mathiasbynens replied:
Could also go with `/'([^']*)'/` (left to right: match `'`, followed by any character that is not `'` 0 or more times, followed by `'`). If you're always expecting at least one character between the quotes, use `+` instead of `*`.

pipobscure replied:
The ? makes the * non-greedy, so it is needed. Also this regex ignores any backslashed quotes. Is that good enough?

How about something like

/‘((?:\\’|[^’])*?)’/

(Apologies for trolling. Could not resist 😀 )

mathiasbynens replied:
I see. I was missing that the full error message is something like

```
The requested module 'foo.mjs' does not provide an export named 'default'
```

(note the second pair of quotes at the end) so non-greediness is indeed needed. TL;DR ignore my first comment :)

MylesBorins replied:
your first suggestion actually doesn't work 😅. It matches between first and last quote in the entire string (there are a couple). The second suggestion works but I'm unsure the advantage

MylesBorins replied:
@pipobscure AFAIK we will never get escaped quotes... I'd prefer to keep it small and readable.

jkrems replied:
Is it possible to decorate the error with proper properties for those values? Or would that blow up this PR too much? It feels awkward if we are the ones burying the value in a string and then also are the ones using a regex to get it back.

MylesBorins commented:
@guybedford getting the specific named export that was attempted to be imported would be a bit more work digging into the internals (is a hidden part of the error message). Happy to include the solution regarding the default but was unsure how verbose to be

ljharb commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -92,6 +93,21 @@ class ModuleJob {
     } catch (e) {
+      if (e.message.includes('does not provide an export named')) {
+        const rootURL = this.module.url;
+        const childSpecifier = e.message.match(/'(.*?)'/)[1];
```
```suggestion
        const childSpecifier = StringPrototypeMatch(e.message, /'(.*?)'/)[1];
```

ljharb commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -92,6 +93,21 @@ class ModuleJob {
+          e.message = `The requested module '${childSpecifier}' is expected ` +
+            'to be of type CommonJS which does not support named exports. ' +
+            'CommonJS modules can be imported by importing the default.';
+          const newStack = e.stack.split('\n');
```
```suggestion
          const newStack = StringPrototypeSplit(e.stack, '\n');
```

ljharb commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -92,6 +93,21 @@ class ModuleJob {
+            'CommonJS modules can be imported by importing the default.';
+          const newStack = e.stack.split('\n');
+          newStack[0] = `Error: ${e.message}`;
+          e.stack = newStack.join('\n');
```
```suggestion
          e.stack = ArrayPrototypeJoin(newStack, '\n');
```

addaleax commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -92,6 +93,21 @@ class ModuleJob {
+        const childURL = new URL(childSpecifier, rootURL);
+        const format = await this.loader.getFormat(childURL.href);
+        if (format === 'commonjs') {
+          e.message = `The requested module '${childSpecifier}' is expected ` +
+            'to be of type CommonJS which does not support named exports. ' +
+            'CommonJS modules can be imported by importing the default.';
```
I feel like it’s still worth mentioning which name was attempted to be loaded, and I’m afraid that “can be imported by importing the default” is a bit technical here (i.e. it would be nice to have a more concrete solution in the error message, even if it’s just an example like ``CommonJS modules can be imported through `import mod from 'module-name'` ``)

devsnek replied:
```suggestion
          e.message = `The requested module '${childSpecifier}' is expected ` +
            'to be of type CommonJS, which only provides default exports.';
```

```suggestion
          e.message = `The requested module '${childSpecifier}' is expected ` +
            'to be of type CommonJS, which only provides a default export.';
```

GeoffreyBooth commented:
> Ideally the error message could contain the exact context and solution, eg something like:

```
The requested module 'pkg' is expected to be of type CommonJS which does not
support named exports. CommonJS modules can be imported by importing the
default, for example: import pkg from 'pkg'; const { name } = pkg;
```

addaleax commented:
@GeoffreyBooth … jinx ;)

MylesBorins commented:
Ok... I've accepted the proposed changes and made another stab at the text. Currently it uses a generic module name and named imports... We should be able to make the error message include the explicit named imports that were used but I'll need to dig in a bit how the message is being decorated with the arrow to do this.

addaleax commented on lib/internal/util.js:
```diff
@@ -101,13 +104,29 @@ function deprecate(fn, msg, code) {
+      'to be of type CommonJS, which does not support named exports. ' +
+      'CommonJS modules can be imported by importing the default export. ' +
+      `The following code will work: const ${namedImports} = ` +
+      `require('${childSpecifier}');`;
```
But `require` doesn’t work in ESM, does it?

MylesBorins replied:
fixed, doing too many things at once LOL

MylesBorins commented:
ok... it is gross but it works 🎉.

Added options argument to the decorate error function and utilized the existing code that grabs the failing import statement. Errors will now point out the error and offer alternative working code.

ljharb commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -92,7 +94,18 @@ class ModuleJob {
-      decorateErrorStack(e);
+      if (e.message.includes('does not provide an export named')) {
+        const rootURL = this.module.url;
+        const childSpecifier = StringPrototypeMatch(e.message, /'(.*?)'/)[1];
```
what happens if this doesn't actually match? couldn't a module have `throw 'does not provide an export named'` in it, and that would break this code? or is this prior to actual user ESM being evaluated?

ljharb commented on lib/internal/util.js:
```diff
@@ -101,13 +104,30 @@ function deprecate(fn, msg, code) {
+    const {
+      childSpecifier
+    } = options;
+    const importStatement = arrow.split('\n')[1];
```
```suggestion
    const importStatement = StringPrototypeSplit(arrow, '\n')[1];
```

benjamingr commented:
I'm not up to date enough in that space right now to approve IMO but changes LGTM.

I just want to make sure that semverness is taken into account and that if modules are stable we make sure to choose the appropriate semverness for an error message change.

MylesBorins commented:
@benjamingr ESM is still marked experimental in the docs and changes are not subject to Semver

mafintosh commented:
@MylesBorins thanks for this! much appreciated

ljharb commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -92,7 +94,22 @@ class ModuleJob {
       }
     } catch (e) {
-      decorateErrorStack(e);
+      if (e.message.includes('does not provide an export named')) {
```
```suggestion
      if (StringPrototypeIncludes(e.message, 'does not provide an export named')) {
```

ljharb commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -92,7 +94,22 @@ class ModuleJob {
+        const rootURL = this.module.url;
+        const childSpecifier = StringPrototypeMatch(
+          e.message,
+          // Match the first string in either single or double quotes
```
this is still not guaranteed to match; the if check doesn’t ensure there’s quotes. If it’s null, this would crash on the [1].

MylesBorins replied:
We can check to ensure but I'm 99% sure this error cannot be reached except when the specifier fails. Will write a bunch of tests though to ensure my assumptions and update accordingly

ljharb replied:
fair enough; it still seems safer to me to use error codes to pivot, and to store this information as a property on the error rather than parsing it out of the error message

MylesBorins commented:
This is catching a V8 error, not an node.js error. I can double check but I
don't believe these errors have error.code

On Wed, May 6, 2020 at 5:18 PM Jordan Harband <notifications@github.com>
wrote:

> *@ljharb* commented on this pull request.
> ------------------------------
>
> In lib/internal/modules/esm/module_job.js
> <https://github.com/nodejs/node/pull/33256#discussion_r421098237>:
>
> > @@ -92,7 +94,22 @@ class ModuleJob {
>          this.module.instantiate();
>        }
>      } catch (e) {
> -      decorateErrorStack(e);
> +      if (e.message.includes('does not provide an export named')) {
> +        const rootURL = this.module.url;
> +        const childSpecifier = StringPrototypeMatch(
> +          e.message,
> +          // Match the first string in either single or double quotes
>
> fair enough; it still seems safer to me to use error codes to pivot, and
> to store this information as a property on the error rather than parsing it
> out of the error message
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/33256#discussion_r421098237>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADZYVZE2I67XE62JGFUFMLRQHHYTANCNFSM4MZ4EM7Q>
> .
>

GeoffreyBooth commented on lib/internal/util.js:
```diff
@@ -101,13 +104,30 @@ function deprecate(fn, msg, code) {
+    err.message = `The requested module '${childSpecifier}' is expected ` +
+      'to be of type CommonJS, which does not support named exports. ' +
+      'CommonJS modules can be imported by importing the default export. ' +
+      'The following code will work:\n' +
+      `import pkg from '${childSpecifier}';\n` +
+      `const ${namedImports} = pkg;`;
```
```suggestion
      'For example:\n' +
      `import pkg from '${childSpecifier}';\n` +
      `const ${namedImports} = pkg;`;
```

I know this seems like a trivial change, but the difference is that you can't actually guarantee that 'the following code will work' because you don't know if `pkg` or the names that are being imported are possibly `const`-declared elsewhere in the file.

It would be nice to use the name of the package instead of `pkg`, if you could find an easy way to convert any string into a valid JS identifier (to handle package names like `@babel/babel-core`).

Also shouldn't `${namedImports}` be something like `{ ${namedImports.split(', ')} }`?

MylesBorins replied:
namedImports is a string that is regex'd out of the failing message, not a list of imports. So it just works ™️

I don't think we should use the package name because this error can come from both a bare + non bare specifier

GeoffreyBooth commented on lib/internal/util.js:
```diff
@@ -120,7 +120,7 @@ function decorateErrorStack(err, options) {
-      'The following code will work:\n' +
+      'For example:\n' +
       `import pkg from '${childSpecifier}';\n` +
       `const ${namedImports} = pkg;`;
```
Can you paste an example of the output when you test this?

Are you going to add a test?

MylesBorins replied:
Working on tests right now, will push something before I go to bed

MylesBorins commented:
PTAL I've added tests and refactored pretty significantly including documentation in the code that gets a bit funky.

I do not have tests for specifiers that have escaped quotes as my various manual tests seem to imply it is impossible to import esmodules in Node.js with escaped quotes in the specifier. If my assumption here is wrong someone please let me know.

targos commented:
> I do not have tests for specifiers that have escaped quotes as my various manual tests seem to imply it is impossible to import esmodules in Node.js with escaped quotes in the specifier. If my assumption here is wrong someone please let me know.

Not sure what you mean. This works:

```
import { value } from './a\'b.js';
```

ljharb commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -92,7 +95,38 @@ class ModuleJob {
+          /(?:'|")(.*?)(?:'|")/
+        )[1];
+        let childFileURL;
+        if (childSpecifier.startsWith('.') || childSpecifier.startsWith('/')) {
```
```suggestion
        if (StringPrototypeStartsWith(childSpecifier, '.') || StringPrototypeStartsWith(childSpecifier, '/')) {
```

ljharb commented on lib/internal/util.js:
```diff
@@ -101,13 +104,30 @@ function deprecate(fn, msg, code) {
+      childSpecifier
+    } = options;
+    const importStatement = StringPrototypeSplit(arrow, '\n')[1];
+    const namedImports = StringPrototypeMatch(importStatement, /{.*}/)[0];
```
```suggestion
    const namedImports = StringPrototypeMatch(importStatement, /{.*}/)?.[0];
```
this would avoid a crash if the match is null, even if you're sure it could never happen :-)

targos replied:
We wouldn't be able to do this in v12.x

MylesBorins replied:
I really don't think we should be adding this additional check if it is not needed.

MylesBorins commented:
OK... so it looks like escaped strings can work as specifiers... and it also looks like V8 normalizes this before printing a message... so independent of if single or double quotes have been used, if quote inside the string are single / double or escaped, the message will always include only single quote 😅

```
SyntaxError: The requested module './oh'no.cjs' does not provide an export named 'value'
```

I've gone ahead and removed the use of a regular expression and am instead manually separating the string based on knowing the shape it will be (this is hardcoded into V8). If this changes internally in V8 all the tests will blow up.

MylesBorins commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -92,7 +97,39 @@ class ModuleJob {
+          // If the specifier is a relative or absolute path make a new
+          // URL object to resolve the file url. This will not work for
+          // bare specifiers.
+          childFileURL = (new URL(childSpecifier, this.module.url)).href;
```
This doesn't get the accurate childFileURL... as it is relative to `this.module.url` which is the entry point to the graph, not the specific parents that imported the child. I've shown a bunch of cases where this is "good enough", but with a couple layers of indirection and nested package.json I think it would be possible to inaccurately resolve the format on line 125.

Calling it a night for now but we likely need to make a test for this. The decorated error has the appropriate parent file url... so we may just have to refactor the order of operations and where the error gets decorated to ensure we appropriately resolve the format.

MylesBorins commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -92,7 +97,39 @@ class ModuleJob {
+          // imported modules. This edge case is highly unlikely, but
+          // probably best to test for it.
+          childFileURL =
+            await this.loader.resolve(childSpecifier, this.module.url);
```
Same as above... if there are nested node_modules folders and duplicated entries it might be possible to resolve the wrong bare import.

The reason we can't use `this.loader.resolve` for the above case is that it checks for the existence of the file when resolving. So in theory once we get the correct base url here we should be able to remove all the comments and the if / else statement and use this.loader.resolve for any child specifier.

MylesBorins commented:
OK... so this was a wild ride but I've fixed all the things. I didn't write a test for the indirection sub package.json edge case I was thinking could be an issue... but all child modules, including bare imports, are 100% being resolved relative to the file that the import statement was contained in 🎉. 

This should cover all cases of escape characters, although might give inaccurate results depending on the string we get from V8.

ljharb commented on test/fixtures/es-modules/package-cjs-named-error/json-hack/fail.js:
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  comeOn: 'fhqwhgads'
```
everybody to the limit

DerekNonGeneric replied:
Maybe `foo` would be more appropriate here?

ljharb replied:
i don't see why anything could possibly be more appropriate than this particular reference

ljharb commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -93,6 +97,28 @@ class ModuleJob {
       decorateErrorStack(e);
+      if (StringPrototypeIncludes(e.message,
+                                  ' does not provide an export named')) {
+        const [ parentFileUrl, importStatement ] =
```
note that using array destructuring here means that `delete Array.prototype[Symbol.iterator]` will break this code. it'd be safer to use explicit index property access (but this is a big problem in node core overall, so nbd if you want to leave it here)

ljharb commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -1,10 +1,14 @@
   SafePromise,
+  StringPrototypeIncludes,
+  StringPrototypeMatch,
+  StringPrototypeSplit
```
is there a reason node's eslint doesn't require trailing object commas?

DerekNonGeneric replied:
It doesn't seem intentional.

https://github.com/nodejs/node/blob/06cdf08b337a2aef127f18f2f6ffab74aa7a18f6/.eslintrc.js#L83

targos replied:
I think the reason is that it would require changing too many lines at once to enable it.

targos commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -1,10 +1,14 @@
   SafePromise,
+  StringPrototypeIncludes,
+  StringPrototypeMatch,
+  StringPrototypeSplit
```
```suggestion
  StringPrototypeSplit,
```

MylesBorins commented:
CI is green. Going to land this EOD unless there are any blockers

MylesBorins commented:
Landed in 2c3c9f5a3f6d
